src/qt/veil/miningwidget.cpp

```diff
@@ -0,0 +1,147 @@
+    mineOn = GenerateActive();
+
+    ui->cmbAlgoSelect->addItems({"RandomX","ProgPow","SHA256D"});
+    ui->cmbAlgoSelect->setCurrentIndex(GetMiningAlgorithm());
+
+    ui->lblMaxThreadsAvailable->setText(QString::number(GetNumCores() - 1));
+    
+
+    connect(ui->btnUpdateAlgo, SIGNAL(clicked()), this, SLOT(onUpdateAlgorithm()));
+    connect(ui->btnAllThreads, SIGNAL(clicked()), this, SLOT(onUseMaxThreads()));
+    connect(ui->btnActiveMine, SIGNAL(clicked()), this, SLOT(onToggleMiningActive()));
+
+}
+
+void MiningWidget::setMineActiveTxt(bool mineActive){
+    ui->lblMineActive->setText(mineActive ? "Mining Active" : "Mining Stopped");
+    ui->lblMineActive->setStyleSheet(mineActive ? "QLabel{color:green;}" : "QLabel{color:red;}");
+    ui->btnActiveMine->setText(mineActive ? "Stop Mining" : "Start Mining");
+}
+
+void MiningWidget::setWalletModel(WalletModel *model){
+    this->walletModel = model;
+
+    connect(model, SIGNAL(updateMiningFields()), this, SLOT(updateMiningFields()));
+}
+
+MiningWidget::~MiningWidget()
+{
+    delete ui;
+}
+
+void MiningWidget::onUpdateAlgorithm() {
+    if (mineOn) {
+        openToastDialog(QString::fromStdString("Mining must be stopped to change algorithm!"), mainWindow->getGUI());
+    } else {
+        currentMiningAlgo = ui->cmbAlgoSelect->currentIndex(); 
+        std::string algoStr = GetMiningType(currentMiningAlgo, false, false);
+        bool setAlgoResult = SetMiningAlgorithm(algoStr); 
+
+        if (setAlgoResult) {
+            openToastDialog(QString::fromStdString("Aglorithm Switch Success!"), mainWindow->getGUI());
+            ui->lblCurrentAlgo->setText(QString::fromStdString(algoStr));
+        } else {
+            openToastDialog(QString::fromStdString("Algorithm Switch Failed!"), mainWindow->getGUI());
+        }
+
+        setThreadSelectionValues(currentMiningAlgo);
+    }
+}
+
+void MiningWidget::onToggleMiningActive() {
+
+    // need to define "coinbaseScript"
+    std::shared_ptr<CReserveScript> coinbase_script;
+    walletModel->wallet().getWalletPointer()->GetScriptForMining(coinbase_script);
+
+    //throw an error if no script was provided
+    if (!coinbase_script || coinbase_script->reserveScript.empty()) {
+       openToastDialog("No coinbase script available", mainWindow->getGUI());
+       return;
+    }
+
+    int nAlgo = GetMiningAlgorithm();
+    int nCores = GetNumCores();
+    int nThreads = ui->numThreads->text().toInt();
+
+    if (!mineOn) {
+        if (nThreads > (nCores - 1)) {
+            openToastDialog("Too many threads", mainWindow->getGUI());
+            return;
+        }
+            
+        if ((nAlgo == MINE_SHA256D) && (nThreads > (nCores - 1)))
+            openToastDialog(QString::fromStdString("SHA256D limited to " + (nCores-1)), mainWindow->getGUI());
+
+        if ((nAlgo == MINE_RANDOMX) && (nThreads < 4)) {
+            openToastDialog(QString::fromStdString("RandomX must be at least 4 threads"), mainWindow->getGUI());
```

Comment:
Is there some way to indicate that this message doesn't prevent mining but changes the threads and starts anyway? Actually, scratch that. I would prefer the GUI enforce that during selection instead of on enabling mining--it wouldn't have to be a pop-up there. And I see below you already have the range being set for randomx.

Reply:
You pretty much hit it on the head.  The GUI is the first line of defense, performing a range check prior to enabling.  This check is for redundancy (in case someone finds a way to get a lower number).